### PR TITLE
tunnel: allow `remote` and `API` to accept `--secrets`

### PR DESCRIPTION
**Pull request by flouthoc**
Following commit makes sure that `build` api can accept external
secret and allows currently `NOOP` `podman-remote build -t tag
--secret id=mysecret,src=/path/on/host` to become functional.

#### TLDR:
Adds support for 
`podman-remote build -t tag
--secret id=mysecret,src=/path/on/host .`

---

Just like `docker` following api is a hidden field and only exposed to
`podman-remote` but could document it if it needs exposed on `swagger`.

Closes: https://github.com/containers/podman/issues/12415

### CODE REVIEW

**Comment by flouthoc** [comment]
Solves: https://github.com/containers/podman/issues/12415#issuecomment-975802180

**Review by vrothberg — COMMENTED** [review]
> [NO NEW TESTS NEEDED]

Please add tests.  Skipping adding tests is something we must discourage, and I've reminded on multiple occasions that we must explain why if there is really no way to test a change.

Changes LGTM

**Comment by flouthoc** [comment]
@vrothberg Sure I am just waiting for discussion to close on something will add as soon as this is ready to be merged. Marking this as `draft` for now. Since this still does not solves the actual problem.

**Comment by flouthoc** [comment]
@vrothberg We could copy secrets over to `remote VM` but not sure if its the right thing to do. I think we should do a bind mount hence this would still not solve cases like `podman-remote build -t tag --secret id=mysecret,src=/path/on/host .`

**Comment by rhatdan** [comment]
It is the right thing to do, copy the secret to the remote machine, but destroy it when the container is removed.  Secrets are secret to the container, IE they do not get committed to the image.  They are not secrets between the client and server.  You might want to clarify this in the man pages, but they need to be copied.  Mounting will not work if you are actually using a remote system as opposed to a VM.

**Comment by flouthoc** [comment]
@rhatdan Sure makes sense. I was also worried about this difference between `mac` and `podman-remote server`. Copy should work for both of these cases. I'll amend my PR.

File: pkg/bindings/images/build.go [hunk 1]
```diff
@@ -384,6 +378,59 @@ func Build(ctx context.Context, containerFiles []string, options entities.BuildO
...
+							// read specified secret into a tmp file
+							// move tmp file to tar and change secret source to relative tmp file
+							tmpSecretFile, err := ioutil.TempFile(options.ContextDirectory, "podman-build-secret")
+							if err != nil {
+								return nil, err
+							}
+							defer os.Remove(tmpSecretFile.Name()) // clean up
+							defer tmpSecretFile.Close()
+							srcSecretFile, err := os.Open(arr[1])
+							defer srcSecretFile.Close()
```
Comment (vrothberg) [inline pkg/bindings/images/build.go:404]: The Close should be moved _after_ the error check.

File: test/e2e/build_test.go [hunk 2]
```diff
@@ -59,6 +59,18 @@ var _ = Describe("Podman build", func() {
 		Expect(session).Should(Exit(0))
 	})
 
+	// following test must trigger for both remote and non-remote
+	It("podman build with a secret from file", func() {
+		session := podmanTest.Podman([]string{"build", "-f", "build/Dockerfile.with-secret", "-t", "secret-test", "--secret", "id=mysecret,src=build/secret.txt", "build/"})
```
Comment (vrothberg) [inline test/e2e/build_test.go:64]: Can we add (another) test with more than 1 secret?

**Comment by vrothberg** [comment]
@ashley-cui PTAL

**Comment by flouthoc** [comment]
@vrothberg thanks resolved in latest commit.

File: test/e2e/build_test.go (hunk 2 above)
Comment (flouthoc) [reply test/e2e/build_test.go:63]: added in latest commit.

**Review by ashley-cui — COMMENTED** [review]
LGTM, thanks!

**Comment by mheon** [comment]
Should we consider any additional security here, given this is secret data? I'm a little concerned that they might be included in built images unintentionally if an image does something like `COPY *`

**Comment by rhatdan** [comment]
@flouthoc please test out @mheon possiblity.  Basically try a Containerfile with

from fedor
copy  * /test/

And then run a test like

podman --remote -f /LOCATIONOF/Containerfile --secret ... /NOTLOCATIONOFContainerfile

Then make sure the secret and Containerfile do not end up i the image.

**Comment by flouthoc** [comment]
@mheon @rhatdan I tested this with `docker` and `buildkit` : `COPY * /` is a user driven command even docker copies everything to working directory from current context. If build secret file is in current context and user wants to run `COPY * /` all files will be copied over since that is what user wants.

I guess we could workaround that by adding extra `.dockerignore` if we want but that would a different behavior though.

**Comment by mheon** [comment]
But does Buildkit move all secrets into the build context, if they weren't already? Because it seems like that's what this code is doing.

**Comment by flouthoc** [comment]
@mheon I haven't tried it on `mac` but i think it should not add secrets to context if they not part of it by default cause they must be `binding` files instead of moving it to context.

I guess we could have a similar thing by checking if secret was not part of current context then add it to `.dockerignore` that should be able to achieve us a similar functionality.

**Comment by flouthoc** [comment]
> @flouthoc please test out @mheon possiblity. Basically try a Containerfile with
> 
> from fedor copy * /test/
> 
> And then run a test like
> 
> podman --remote -f /LOCATIONOF/Containerfile --secret ... /NOTLOCATIONOFContainerfile
> 
> Then make sure the secret and Containerfile do not end up i the image.

@mheon @rhatdan Following case is handled in a new commit also added a new test to verify the same.

**Comment by flouthoc** [comment]
@rhatdan @mheon PTAL. New test verifies the concern.

File: pkg/api/handlers/compat/images_build.go [hunk 3]
```diff
@@ -239,6 +240,49 @@ func BuildImage(w http.ResponseWriter, r *http.Request) {
...
+				for _, token := range secretOpt {
+					arr := strings.SplitN(token, "=", 2)
+					if len(arr) > 1 {
+						if arr[0] == "src" {
+							/* move secret away from contextDir */
+							/* to make sure we dont accidentally commit temporary secrets to image*/
+							builderDirectory, _ := filepath.Split(contextDirectory)
+							// following path is outside build context
+							newSecretPath := filepath.Join(builderDirectory, arr[1])
+							oldSecretPath := filepath.Join(contextDirectory, arr[1])
```
Comment (mheon) [inline pkg/api/handlers/compat/images_build.go:267]: Do we clean this directory out anywhere, remove the files inside it? Is it shared between multiple builds, potentially?

Comment (flouthoc) [reply pkg/api/handlers/compat/images_build.go:267]: @mheon Yes build directory is automatically cleaned up and no its a new temp directory for each build.

**Comment by mheon** [comment]
LGTM

**Comment by flouthoc** [comment]
@rhatdan PTAL

**Pull request merged**